### ISSUE TITLE
CA2012 ValueTask Consumption Improvements

### ISSUE THREAD

**Issue by CoolCoderSuper** [dotnet/roslyn-analyzers#7461]
### Analyzer

**Diagnostic ID**: [CA2012](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/CA2012)

### Describe the improvement

Sometimes method other than the built in ones are valid consumptions.

### Describe suggestions on how to achieve the rule

Add editor config option similar to other analyzers

### Additional context

```vb
Public Async Function Test() As Task
    Await ValueTask.KeepContext()
End Function

Public Module AsyncExtensions
    <Extension>
    Public Function KeepContext(a As ValueTask) As ConfiguredValueTaskAwaitable
        Return a.ConfigureAwait(True)
    End Function

    <Extension>
    Public Function FreeContext(a As ValueTask) As ConfiguredValueTaskAwaitable
        Return a.ConfigureAwait(False)
    End Function
End Module
```
dotnet_code_quality.additional_valid_valuetask_consumption=KeepContext|FreeContext

**Comment by sharwell** [comment]
> Sometimes method other than the built in ones are valid consumptions.

Can you provide a bit more context here? In the sample above, it doesn't appear any new valid consumptions are added.

**Comment by CoolCoderSuper** [comment]
In the example the FreeContext method is a wrapper around ConfiureAwait but calling it on a value task triggers the warning even it is fine to use that method.
The idea I had in mind was in some cases I know methods to be safe consumers.
I just copied what I had seen in other spots so it might not be the best idea.

But a DiagnosticSuppressor would still need configuration not? Or would this be in my own analyzer.

**Comment by CoolCoderSuper** [comment]
Maybe I don't completely understand the best practice either
